client/src/leap/soledad/client/http_target.py

```diff
@@ -0,0 +1,570 @@
+        else:
+
+            def _maybe_encrypt_doc_inline(doc_json):
+                if doc_json is None:
+                    # the document is not marked as tombstone, but we got
+                    # nothing from the sync db. As it is not encrypted
+                    # yet, we force inline encryption.
+                    return encrypt_doc(self._crypto, doc)
+                return doc_json
+
+            d = self._sync_enc_pool.get_encrypted_doc(doc.doc_id, doc.rev)
+            d.addCallback(_maybe_encrypt_doc_inline)
+        return d
+    
+    #
+    # methods to receive doc
+    #
+
+    @defer.inlineCallbacks
+    def _receive_docs(self, last_known_generation, last_known_trans_id,
+                      return_doc_cb, ensure_callback, sync_id,
+                      defer_decryption):
+        # we keep a reference to the callback in case we defer the decryption
+        self._return_doc_cb = return_doc_cb
+        self._queue_for_decrypt = defer_decryption \
+            and self._sync_db is not None
+
+        new_generation = last_known_generation
+        new_transaction_id = last_known_trans_id
+
+        if self._queue_for_decrypt:
+            logger.debug(
+                "Soledad sync: will queue received docs for decrypting.")
+
+        if defer_decryption:
+            self._setup_sync_decr_pool()
+
+        headers = self._auth_header.copy()
+        headers.update({'content-type': 'application/x-soledad-sync-get'})
+
+        # maybe get one doc
+        d = self._receive_one_doc(
+            headers, last_known_generation, last_known_trans_id,
+            sync_id, 0)
+        d.addCallback(partial(self._insert_received_doc, 1, 1))
+        number_of_changes, ngen, ntrans = yield d
+
+        if defer_decryption:
+            self._sync_decr_pool.set_docs_to_process(
+                number_of_changes)
+        idx = 1
+
+        # maybe get more documents
+        deferreds = []
+        while idx < number_of_changes:
+            d = self._receive_one_doc(
+                headers, last_known_generation,
+                last_known_trans_id, sync_id, idx)
+            d.addCallback(
+                partial(
+                    self._insert_received_doc,
+                    idx + 1,
+                    number_of_changes))
+            deferreds.append(d)
+            idx += 1
+        results = yield defer.gatherResults(deferreds)
+
+        # get genration and transaction id of target after insertions
+        if deferreds:
+            _, new_generation, new_transaction_id = results.pop()
+
+        # get current target gen and trans id in case no documents were
+        def _shutdown_and_finish(res):
+            self._sync_decr_pool.close()
+            return new_generation, new_transaction_id
+
+        d = defer.Deferred()
+        d.addCallback(_shutdown_and_finish)
+
+        def _wait_or_finish():
+            if not self._sync_decr_pool.has_finished():
+                reactor.callLater(
```

Comment:
in my tests the sync_decr_pool is never able to finish, I'm debugging what could be the cause.